rpm/macros.d/macros.go-srpm

```diff
@@ -168,6 +168,6 @@ if (rpm.expand("%{?-i}") ~= "") then
   rpm.expand("%{echo:  gourl:           %{?gourl}}")
   rpm.expand("%{echo:  gosource:        %{?gosource}}")
 end}
-BuildRequires: compiler(go-compiler)
+BuildRequires: go-rpm-macros
```

Comment:
Please, stick to the original BuildRequires

Reply:
I added the rationale in comment as you wanted

Reply:
But anyway, even if you do insist on separate src.rpms for the macros (with all the breakage risks that entails) that does not change the dependency logic between the packages themselves

rpm dnf and yum work on binary packages, they don't care at all what is the origin src.rpm of a noarch or arch rpm (that's what makes a lot or relayouts possible)

Reply:
I don't expect the content of the `macros.go-srpm` to change very often, if at all. We should update it only if really necessary. In general, the `macros.go-srpm` should be independent of what we ship in the `go-compilers` package.

Reply:
The more I work with rpm and other software deployment technologies, the more I am convinced there is a *strong* relationship between respecting project boundaries in components and getting working results. Every single deployment tech I've seem that tried to ignore project boundaries (call it vendoring, bundling or whatever else) ended up with no end of problems (rotten code, legal problems, incompatible components, awful fix propagation timescales).

So, even if rpm is powerful enough to mashup several projects in one src.rpm, or spilling the same project in several src.rpm, doing so is a terrible idea and usually end up badly. It's a shortcut, and ends up like most shortcuts do, with people saying 'I though it would save me efforts' or 'I thought it was not necessary to do' at the time.

Nevertheless, once again, dnf and yum work on produced arch/noarch packages and do not care what src.rpm is the source of a particular package, so doing deps correctly this way is independent of producing the go-rpm-macros and go-srpm-macros from the same src.rpm or from two separate src.rpms.

Reply:
You seems to me to be in disagreement with yourself on this one. Could you please rephrase you  point? As I'm not really sure if you are opposing general rpm packaging or supporting it.

There is no issue in the boundaries as you approach is actually violating them as you are mixing matching rpm macros with srpm macros, those are two totally different problem solving fields and one is building in top of the other (rpm macros on top of the srpm).

Reply:
```
upstream project    spec/src.rpm             deployment(sub-)packages

                                             ┌─╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌─╌╌╌╌╌╌╌╌╌╌╌┐
go-macros──────────▶go-macros─────────────┬─▶┊go-srpm-macros◀─────────────────────────────────────────┐           ┊ Actual buildroot content
  Version: X          Version: X          │  ┊  Version: X                                            │           ┊ Dangerous files for other packages forbidden
  Macro files                             │  ┊  Architecture: noarch                                  │           ┊ Requiring packages outside the buildroot forbidden
  and shell                               │  ┊  Files:                                                │           ┊ Being required for version-lock is allowed
  wrappers                                │  ┊   /usr/lib/rpm/macros.d/macros.go-rpm      R: go-srpm-macros = X   ┊
                                          │  └─╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┘
                                          │                                                           │
                                          ├──▶go-rpm-macros───────────────────────────────────────────┼───────────┐
                                          │     Version: X                                            │           │
                                          │     Architecture: noarch                     R: compiler(go-compiler) │
                                          │     Files:                                                │           │
                                          │       /usr/bin/go-rpm-integration                         │       R: golist
                                          │       /usr/lib/rpm/fileattrs/go.attr                      │           │
                                          │       /usr/lib/rpm/fileattrs/gobundled.attr               │           │
                                          │       /usr/lib/rpm/fileattrs/gosymlink.attr               │           │
                                          │       /usr/lib/rpm/gobundled.prov                         │           │
                                          │       /usr/lib/rpm/gosymlink.deps                         │           │
                                          │       /usr/lib/rpm/macros.d/macros.go-rpm                 │           │
                                          │                                                           │           │
                                          ├──▶go-compiler-golang◀─────────────────────────────────────┤           │
                                          │     Version: X                                            │           │
                                          │     Architecture: arch                                    │           │
                                          │     Provides:                                             │           │
                                          │       compiler(go-compiler) = 2                           │           │
                                          │       compiler(golang)                                    │           │
                                          │     Files:                                                │           │
                                          │       /usr/lib/rpm/macros.d/macros.go-compiler-golang     │           │
                                          │                                                           │           │
                                          └──▶go-compiler-gcc◀────────────────────────────────────────┘           │
                                                Version: X                                                        │
                                                Architecture: arch                                                │
                                                Provides:                                                         │
                                                  compiler(go-compiler) = 1                                       │
                                                  compiler(gcc-go)                                                │
                                                Files:                                                            │
                                                  /usr/lib/rpm/macros.d/macros.go-compiler-gcc                    │
                                                                                                                  │
symbols-extractor──▶go-symbols-extractor─────▶go-symbols-extractor◀───────────────────────────────────────────────┘
  Version: Y          Version: Y                Version: Y
  Go code                                       Architecture: arch
                                                Provides: golist
                                                Files:
                                                  /usr/bin/golist

└──────────────────────────────────────┘      └───────────────────────────────────────────────────────────────────┘
 What exists from a package maintenance         What exists for the software installer, on the target systems and
 point of view. Direct 1:1 mapping is           for installing the buildroot. Does not care what is the source
 desired to limit maintenance costs and         package from which those packages were built
 making sure all the parts of a project
 are made available simultaneously
 during a Fedora compose
```

Reply:
And if you are absolutely convinced src.rpm layout has any influence on the buildroot (it hasn’t) and want to break the 1:1 property which is desirable to limit maintenance costs
 
```
upstream project    spec/src.rpm             deployment(sub-)packages

                                             ┌─╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌─╌╌╌╌╌╌╌╌╌╌╌┐
go-macros────────┬─▶go-srpm-macros──────────▶┊go-srpm-macros◀─────────────────────────────────────────┐           ┊ Actual buildroot content
  Version: X     │    Version: X             ┊  Version: X                                            │           ┊ Dangerous files for other packages forbidden
  Macro files    │                           ┊  Architecture: noarch                                  │           ┊ Requiring packages outside the buildroot forbidden
  and shell      │                           ┊  Files:                                                │           ┊ Being required for version-lock is allowed
  wrappers       │                           ┊   /usr/lib/rpm/macros.d/macros.go-rpm      R: go-srpm-macros = X   ┊
                 │                           └─╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┘
                 │                                                                                    │
                 └─▶go-rpm-macros─────────┬──▶go-rpm-macros───────────────────────────────────────────┼───────────┐
                      Version: X          │     Version: X                                            │           │
                                          │     Architecture: noarch                     R: compiler(go-compiler) │
                                          │     Files:                                                │           │
                                          │       /usr/bin/go-rpm-integration                         │       R: golist
                                          │       /usr/lib/rpm/fileattrs/go.attr                      │           │
                                          │       /usr/lib/rpm/fileattrs/gobundled.attr               │           │
                                          │       /usr/lib/rpm/fileattrs/gosymlink.attr               │           │
                                          │       /usr/lib/rpm/gobundled.prov                         │           │
                                          │       /usr/lib/rpm/gosymlink.deps                         │           │
                                          │       /usr/lib/rpm/macros.d/macros.go-rpm                 │           │
                                          │                                                           │           │
                                          ├──▶go-compiler-golang◀─────────────────────────────────────┤           │
                                          │     Version: X                                            │           │
                                          │     Architecture: arch                                    │           │
                                          │     Provides:                                             │           │
                                          │       compiler(go-compiler) = 2                           │           │
                                          │       compiler(golang)                                    │           │
                                          │     Files:                                                │           │
                                          │       /usr/lib/rpm/macros.d/macros.go-compiler-golang     │           │
                                          │                                                           │           │
                                          └──▶go-compiler-gcc◀────────────────────────────────────────┘           │
                                                Version: X                                                        │
                                                Architecture: arch                                                │
                                                Provides:                                                         │
                                                  compiler(go-compiler) = 1                                       │
                                                  compiler(gcc-go)                                                │
                                                Files:                                                            │
                                                  /usr/lib/rpm/macros.d/macros.go-compiler-gcc                    │
                                                                                                                  │
symbols-extractor──▶go-symbols-extractor─────▶go-symbols-extractor◀───────────────────────────────────────────────┘
  Version: Y          Version: Y                Version: Y
  Go code                                       Architecture: arch
                                                Provides: golist
                                                Files:
                                                  /usr/bin/golist

└──────────────────────────────────────┘      └───────────────────────────────────────────────────────────────────┘
 What exists from a package maintenance         What exists for the software installer, on the target systems and
 point of view. Direct 1:1 mapping is           for installing the buildroot. Does not care what is the source
 desired to limit maintenance costs and         package from which those packages were built
 making sure all the parts of a project
 are made available simultaneously
 during a Fedora compose
```
That’s a more complex layout with direct maintenance drawbacks and unobvious mapping between upstream project names and spec names.

However, regardless of the layout chosen the Requires rules between all the packages are the same and the BuildRequires in %gometa is still go-rpm-macros

Reply:
And point is?(those ascii art pictures are quiet hard to read due to their size and github formatting/line-wrapping)

And I think that you have and error there as upstream is the dist git repository. We are not trying to achieve some theoretical purity as you self have said. And in this case there is no "right" way both are broken in same way. Current setup allows more granulated access to the repositories via commit ACL.

Reply:
The point is that the built package dependency graph is independent of how you choose to break out the srpms